pkg/dns/aws/dns.go

```diff
@@ -190,31 +190,27 @@ func (m *Manager) getLBHostedZone(name string) (string, error) {
 	m.lock.Lock()
 	defer m.lock.Unlock()
 
-	if id, exists := m.lbZones[name]; exists {
-		return id, nil
+	if zoneID, exists := m.lbZones[name]; exists {
+		return zoneID, nil
 	}
 
-	var id string
-	fn := func(resp *elb.DescribeLoadBalancersOutput, lastPage bool) (shouldContinue bool) {
-		for _, lb := range resp.LoadBalancerDescriptions {
-			log.V(0).Info("found load balancer", "name", aws.StringValue(lb.LoadBalancerName), "dns name", aws.StringValue(lb.DNSName), "hosted zone ID", aws.StringValue(lb.CanonicalHostedZoneNameID))
-			if aws.StringValue(lb.CanonicalHostedZoneName) == name {
-				id = aws.StringValue(lb.CanonicalHostedZoneNameID)
-				return false
-			}
-		}
-		return true
-	}
-	err := m.elb.DescribeLoadBalancersPages(&elb.DescribeLoadBalancersInput{}, fn)
+	// An elb name is the first 32 bytes of the elb's dns name.
+	// https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L86-L89
+	lbName := name[:32]
```

Comment:
From the cited code, we know that the name of the load balancer is a function of the UID of the service.  However, `name` is the host name of the service, right? If we had the service's UID, we could go through the steps in the cited code (take the UID, prepend an 'a', elide hyphens, and truncate to 32 characters) to get the LB's name, but we have the host name, not the UID.

Reply:
> However, name is the host name of the service, right?

Name is derived from the first 32-bytes of the service's `uid`. For example a service with `uid: 5b039b2f-72b4-11e9-b976-0a20dbb71566`, has an elb name of `a5b039b2f72b411e9b9760a20dbb7156`. The elb's dns name is equivalent to the service's `hostname`.

Reply:
> Name is derived from the first 32-bytes of the service's `uid`.

Correct.

> The elb's dns name is equivalent to the service's `hostname`.

Correct ([source](https://github.com/kubernetes/cloud-provider-aws/blob/8f6e1e3d2cfa20a0deac088b9c4022d433500369/pkg/cloudprovider/providers/aws/aws.go#L3671-L3676)).

The assumption that I want to validate is that the load balancer's `DNSName` is the same as its `LoadBalancerName`.  The most definitive statement I have found so far in the AWS documentation is the following:

> Each Classic Load Balancer receives a default Domain Name System (DNS) name. This DNS name includes the name of the AWS region in which the load balancer is created. For example, if you create a load balancer named my-loadbalancer in the US West (Oregon) region, your load balancer receives a DNS name such as my-loadbalancer-1234567890.us-west-2.elb.amazonaws.com.

https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/using-domain-names-with-elb.html

I'm hoping for a stronger API commitment than a "for example".

Reply:
Assumption: only the cloud provider knows the opaque platform-specific load balancer identifier, because the cloud provider creates the load balancer.

Is it thus the responsibility of the cloud provider to surface this sort of information to downstream consumers? I'm not sure what form that would take on a `corev1.Service` (conditions? fields?), but without it we're left essentially reverse engineering potentially volatile internal cloud provider naming functions.